An isocrystal story

robertylewis opened the pull request:
After one lunch a week ago: "half a blog post on isocrystals." After a few more lunches and a marathon Zoom call, the second half :smile: 

This could use a close reading by a number theorist. @jcommelin @adamtopaz @riccardobrasca @kbuzzard what did we get wrong?

cc @hrmacbeth @dupuisf

jcommelin commented on posts/formalizing-isocrystals-in-lean.md:
```diff
@@ -0,0 +1,211 @@
+But this is not the full extent of the generalization!  Our number theorist friends here in mathlib told us that we should
+make sure we chose this full generality of *semilinear* maps, maps $f:M \to N$ such that $f(ax)=\sigma(a)f(x)$ for some ring homomorphism $\sigma$ between the scalar rings of the modules $M$ and $N$.  So we and our coauthor Frédéric Dupuis implemented this full generality, and then asked them for an example to illustrate their need for this more abstract definition. This blog post tells the story ...
+
+It turns out that the standard use of semilinear maps in number theory is for *Frobenius-semilinearity*, semilinearity with
+respect to the ring homomorphism of the fraction field of the $p$-typical Witt vectors over a perfect characteristic-$p$ integral domain which is induced by the Frobenius
+automorphism of that domain.  Let's backtrack to catch everyone up...
```
Since this is the "teaser", you could consider saying something like
```suggestion
It turns out that the standard use of semilinear maps in number theory is for *Frobenius-semilinearity*,
semilinearity with respect to ring homomorphisms that are related to the Frobenius homomorphisms $x ↦ x^p$
that show up in the study of finite fields of characteristic $p$.
In our work, this Frobenius homomorphism manifested itself as an automorphism of
the fraction field of the $p$-typical Witt vectors over a perfect characteristic-$p$ integral domain
which is induced by the Frobenius automorphism of that domain.
Let's backtrack to catch everyone up...
```
That way the first reminder on Frobenius is maybe a little bit less intimidating.

faenuccio replied:
Why don't you even prefer "study of algebras over fields of characteristic $p$"? There is no good reason to stick to finite ones, and in particular most of the geometric instances come from infinite F_p-algebras, no?

kbuzzard replied:
If you want to make it even less intimidating, you could simply write "It turns out that one standard use of semilinear maps in number theory is *Frobenius-semilinearity*, semilinearity with respect to the Frobenius endomorphism associated to a prime number p. The Frobenius endomorphism can be viewed as a non-archimedean analogue of complex conjugation, so perhaps one could argue that the two different motivations for formalising semilinear maps boil down to an archimedean and a nonarchimedean avatar of the same underlying idea.

faenuccio replied:
This is also a very nice way of putting it, yes.

robertylewis replied:
I think the stylistic point of this sentence before the teaser was to give a *very* small taste of what's coming, postponing the real details and exposition for later. It's supposed to be intimidating for effect. (Hence "let's backtrack...") So I'm not sure expanding it (beyond making sure it's technically correct) is the right move. Heather, what do you think?

jcommelin commented on posts/formalizing-isocrystals-in-lean.md:
```diff
@@ -0,0 +1,211 @@
+of $\mathbb{Z}/p\mathbb{Z}$.
+
+A ring $R$ of characteristic $p$ has an endomorphism, constructed by sending each element $x$ to $x ^ p$.  $R$ is called *perfect* if this
+endomorphism is an automorphism.  And under mild further conditions ($R$ an integral domain) this automorphism "lifts" <!-- check terminology -->
+to an automorphism of the field of fractions of $\mathbb{W}(R)$, which we will denote $\varphi$.  It's *Frobenius-semilinearity*, i.e. semilinearity with respect $\varphi$, that the number theorists
+want to consider! 
```
If you want, you could say that by functoriality you get a Frobenius automorphism of `W(R)` for every `R`. Then, if `R` is an integral domain, so is `W(R)` and it makes sense to talk about the field of fractions. The automorphism will extend to the field of fractions because that's true for every automorphism of an integral domain.

robertylewis replied:
I'm not totally sure for what audience this will clarify things. My sense is that anyone who gets your two-sentence version will also get our (slightly vaguer) one-sentence version, or do you think that the extra detail would be helpful?

jcommelin commented on posts/formalizing-isocrystals-in-lean.md:
```diff
@@ -0,0 +1,211 @@
+want to consider! 
+
+The fundamental result here is a classification theorem for
+Frobenius-semilinear automorphisms of finite-dimensional vector spaces over $K$, the field of fractions of $\mathbb{W}(k)$, for $k$ an algebraically closed field (and hence a perfect integral domain).  Such an object (a finite-dimensional 
```
Maybe introduce `K` as notation for the field of fractions in the preceding paragraph, when you also introduce `ϕ`?

hrmacbeth replied:
I think it can't quite be done while remaining literally correct, because in the previous paragraph we discuss the fraction field of `𝕎 R`, but in this paragraph we discuss the fraction field of `𝕎 k`, for `k` some algebraically closed field.  So to remain literally correct one would have to refer to `K(R)`, `K(k)`, etc. which seems unwieldy.

jcommelin commented on posts/formalizing-isocrystals-in-lean.md:
```diff
@@ -0,0 +1,211 @@
+
+So here is the answer.  The Witt vectors $\mathbb{W}(k)$ live inside their fraction field $K$ and there is a distinguished multiplicative subgroup $\mathbb{W}(k)^\times$ of $K^*$, the original units of $\mathbb{W}(k)$.  In the representation of $\mathbb{W}(k)$ as a sequence of elements of $k$, the units are [precisely](https://github.com/leanprover-community/mathlib/blob/lean-3.44.1/src/ring_theo ...
+
+The $p$-typical Witt vector $(0,1,0,\ldots)$ is referred to in $\mathbb{W}(k)$ as $p$ (it actually ends up being the sum of $p$ copies of $1=(1,0,0,\ldots)$, under the crazy Witt vector addition).  Multiplication (in the crazy Witt vector sense) by $p$ sends any Witt vector to a Witt vector with $0$ in the first position, and more generally multiplication by $p^m$ for any $m\in \mathbb{N}$ sends  ...
```
I personally always picture Witt vectors as trailing off to the left. Just like with `p`-adic "decimal expansions" of `p`-adic integers.

jcommelin replied:
Note that
```
/-- The composition of Frobenius and Verschiebung is multiplication by `p`. -/
lemma frobenius_verschiebung (x : 𝕎 R) :
  frobenius (verschiebung x) = x * p :=
by { ghost_calc x, ghost_simp [mul_comm] }
```
tells you even more precisely what multiplication by `p` looks like, when `R` has characteristic `p`: you shift all the coefficients, introducing a `0` in the first position, and then you raise every coefficient to the power `p`. (Of course `0^p = 0`, so the first coeff is still `0`.)

faenuccio replied:
Why do you use $^\ast$ to denote units of the field, and $^\times$ for units in the ring?

hrmacbeth replied:
Implemented the two notational suggestions, thanks!  Regarding @jcommelin's comment about explaining what multiplication by `p` looks like, this seems like maybe too much detail for the blog post, but we added a link to that theorem on that sentence.  OK?

jcommelin reviewed (commented):
Great story!

kbuzzard commented:
You know that Frobenius *is* complex conjugation, right? Frobenius is the canonical Galois element for a p-adic place, like complex conjugation is the canonical Galois element for an infinite place.

faenuccio commented on posts/formalizing-isocrystals-in-lean.md:
```diff
@@ -0,0 +1,211 @@
+
+So here is the answer.  The Witt vectors $\mathbb{W}(k)$ live inside their fraction field $K$ and there is a distinguished multiplicative subgroup $\mathbb{W}(k)^\times$ of $K^*$, the original units of $\mathbb{W}(k)$.  In the representation of $\mathbb{W}(k)$ as a sequence of elements of $k$, the units are [precisely](https://github.com/leanprover-community/mathlib/blob/lean-3.44.1/src/ring_theo ...
+
+The $p$-typical Witt vector $(0,1,0,\ldots)$ is referred to in $\mathbb{W}(k)$ as $p$ (it actually ends up being the sum of $p$ copies of $1=(1,0,0,\ldots)$, under the crazy Witt vector addition).  Multiplication (in the crazy Witt vector sense) by $p$ sends any Witt vector to a Witt vector with $0$ in the first position, and more generally multiplication by $p^m$ for any $m\in \mathbb{N}$ sends  ...
```
```suggestion
The $p$-typical Witt vector $(0,1,0,\ldots)$ is referred to in $\mathbb{W}(k)$ as $p$ (it actually ends up being the sum of $p$ copies of $1=(1,0,0,\ldots)$, under the crazy Witt vector addition).  Multiplication (in the crazy Witt vector sense) by $p$ sends any Witt vector to a Witt vector with $0$ in the first position, and more generally multiplication by $p^m$ for any $m\in \mathbb{N}$ sends any Witt vector to a Witt vector with $m$ leading zeros. <!-- double-check how it works --> In fact, every nonzero Witt vector is [of the form](https://github.com/leanprover-community/mathlib/blob/lean-3.44.1/src/ring_theory/witt_vector/discrete_valuation_ring.lean#L117)  $p^m$ for some $m\in\mathbb{N}$ and some $w\in\mathbb{W}(k)^\times$.  The statement for the fraction field $K$ is only slightly more complicated -- every nonzero element of $K$ is of the form $p^m w$ for some $m\in\mathbb{Z}$ and some Witt vector $w\in\mathbb{W}(k)^\times$. 
```

robertylewis replied:
Don't have time to address all the comments right now, but just so it doesn't go unnoticed later, there's a `w` dropped in your suggestion! 

```suggestion
The $p$-typical Witt vector $(0,1,0,\ldots)$ is referred to in $\mathbb{W}(k)$ as $p$ (it actually ends up being the sum of $p$ copies of $1=(1,0,0,\ldots)$, under the crazy Witt vector addition).  Multiplication (in the crazy Witt vector sense) by $p$ sends any Witt vector to a Witt vector with $0$ in the first position, and more generally multiplication by $p^m$ for any $m\in \mathbb{N}$ sends any Witt vector to a Witt vector with $m$ leading zeros. <!-- double-check how it works --> In fact, every nonzero Witt vector is [of the form](https://github.com/leanprover-community/mathlib/blob/lean-3.44.1/src/ring_theory/witt_vector/discrete_valuation_ring.lean#L117)  $p^m w$ for some unique $m\in\mathbb{N}$ and some unique $w\in\mathbb{W}(k)^\times$.  The statement for the fraction field $K$ is only slightly more complicated -- every nonzero element of $K$ is of the form $p^m w$ for some $m\in\mathbb{Z}$ and some Witt vector $w\in\mathbb{W}(k)^\times$. 
```

faenuccio replied:
Not sure I see it, but if you saw it once you will see it later again! ;-)

faenuccio replied:
Right!

faenuccio commented on posts/formalizing-isocrystals-in-lean.md:
```diff
@@ -0,0 +1,211 @@
+The $p$-typical Witt vector $(0,1,0,\ldots)$ is referred to in $\mathbb{W}(k)$ as $p$ (it actually ends up being the sum of $p$ copies of $1=(1,0,0,\ldots)$, under the crazy Witt vector addition).  Multiplication (in the crazy Witt vector sense) by $p$ sends any Witt vector to a Witt vector with $0$ in the first position, and more generally multiplication by $p^m$ for any $m\in \mathbb{N}$ sends  ...
+
+<a name="existence-lemma"></a>
+This subgroup $\mathbb{W}(k)^\times$ turns out to be precisely the things in $K$ which can be expressed as $\varphi(b)/b$ for some $b \in K$ (and it turns out that $b \in \mathbb{W}(k)$ suffices). Make a note of this statement. The rest of the math in this blog post will be devoted to proving this.
```
```suggestion
This subgroup $\mathbb{W}(k)^\times$ turns out to be precisely the things in $K$ which can be expressed as $\varphi(b)/b$ for some $b \in K^\times$ (and it turns out that $b \in \mathbb{W}(k)\setminus {0}$ suffices). Make a note of this statement. The rest of the math in this blog post will be devoted to proving this.
```

faenuccio commented:
> You know that Frobenius _is_ complex conjugation, right? Frobenius is the canonical Galois element for a p-adic place, like complex conjugation is the canonical Galois element for an infinite place.

@kbuzzard Are you sure that you would suggest to write that Frobenius _is_ complex conjugation? This opens the Pandora vase of deciding whether $\mathbb{C}/\mathbb{R}$ is ramified or unramified, and from the CFT perspective it might be considered ramified. I certainly agree that there is a _very strong_ analogy and that one should mention this, but I wonder if you really want to say that it _is_ the conjugation.

kbuzzard commented:
I absolutely agree that figuring out whether C/R is ramified or not is somehow a very confusing question; perhaps this is an indication that one should only be asking questions about ramification for nonarchimedean valuations. But I think that this is not the point here: I am just suggesting an analogy.

What struck me about the post was that it starts by saying "we wanted to unify linear and conjugate-linear maps; then the number theorists told us they wanted semilinear maps and when we asked why, they said that they used Frobenius-linear maps which are some complicated thing" (they're not a complicated thing, they are a lift of the char p ring homomorphism x \mapsto x^p to a char 0 ring homomorphism). What I want to point out is that the writing of this blog post clearly flags up the following interesting (to me) point: the analysts wanted semilinear maps because they wanted conjugate-linear maps, the number theorists wanted semilinear maps because they wanted Frobenius-linear maps, and I'm saying that there's now a great opportunity to drop in a line saying "actually these two things are analogues of each other, because complex conjugation can be thought of as Frobenius at infinity". The observation can just be left at that, but for me it's a striking observation; as a number theorist I "know" that obviously we need semi-linear maps because I see them everywhere in p-adic Hodge theory, but now you come to mention it those semilinear maps are all Frobenius-linear, and the example which motivated the refactor can also be thought of as a Frobenius-linear map.

kbuzzard commented on posts/formalizing-isocrystals-in-lean.md:
```diff
@@ -0,0 +1,211 @@
+
+Back in 2020, one of us (Rob), together with Johan Commelin, [formalized the theory of Witt vectors](https://arxiv.org/abs/2010.02595) in Lean.  The $p$-typical Witt 
+vectors $\mathbb{W}(R)$ over a ring $R$  are, concretely, sequences of elements of $R$, but equipped with a rather wild multiplicative and additive
+structure (dependent on $p$) to make this set into a commutative ring.  The canonical example is the $p$-adic numbers, which are the Witt vectors
```
```suggestion
structure (dependent on $p$) to make this set into a commutative ring.  The canonical example is the $p$-adic integers, which are the Witt vectors
```

kbuzzard commented on posts/formalizing-isocrystals-in-lean.md:
```diff
@@ -0,0 +1,211 @@
+
+A ring $R$ of characteristic $p$ has an endomorphism, constructed by sending each element $x$ to $x ^ p$.  $R$ is called *perfect* if this
+endomorphism is an automorphism.  And under mild further conditions ($R$ an integral domain) this automorphism "lifts" <!-- check terminology -->
+to an automorphism of the field of fractions of $\mathbb{W}(R)$, which we will denote $\varphi$.  It's *Frobenius-semilinearity*, i.e. semilinearity with respect $\varphi$, that the number theorists
```
```suggestion
to an automorphism of the field of fractions of $\mathbb{W}(R)$, which we will denote $\varphi$.  It's *Frobenius-semilinearity*, i.e. semilinearity with respect to $\varphi$, that the number theorists
```

faenuccio commented:
> I absolutely agree that figuring out whether C/R is ramified or not is somehow a very confusing question; perhaps this is an indication that one should only be asking questions about ramification for nonarchimedean valuations. But I think that this is not the point here: I am just suggesting an analogy (which you agree with :-) )
> 
> What struck me about the post was that it starts by saying "we wanted to unify linear and conjugate-linear maps; then the number theorists told us they wanted semilinear maps and when we asked why, they said that they used Frobenius-linear maps which are some complicated thing" (they're not a complicated thing, they are a lift of the char p ring homomorphism x \mapsto x^p to a char 0 ring homomorphism). What I want to point out is that the writing of this blog post clearly flags up the following interesting (to me) point: the analysts wanted semilinear maps because they wanted conjugate-linear maps, the number theorists wanted semilinear maps because they wanted Frobenius-linear maps, and I'm saying that there's now a great opportunity to drop in a line saying "actually these two things are analogues of each other, because complex conjugation can be thought of as Frobenius at infinity". The observation can just be left at that, but for me it's a striking observation; as a number theorist I "know" that obviously we need semi-linear maps because I see them everywhere in p-adic Hodge theory, but now you come to mention it those semilinear maps are all Frobenius-linear, and the example which motivated the refactor can also be thought of as a Frobenius-linear map.

Yes, I agree very much. Perhaps, we can simply say that both complex conjugation and Frobenius are canonical generators of "unique" Galois groups ($\mathbb{C}$ is the unique algebraic extension of $\mathbb{R}$, and $\mathrm{Frac}(W(k))$ is the unique unramified extension of $\mathbb{Q}_p$ of degree $[k:\mathbb{F}_p]$); I certainly agree that this analogy cannot be put under the rug.

eric-wieser commented on posts/formalizing-isocrystals-in-lean.md:
```diff
@@ -0,0 +1,211 @@
+Given the structure of the statement we're working toward, it's not surprising that we'll need to dive into how multiplication on Witt vectors works. In this section we'll prove the
+[following lemma](https://github.com/leanprover-community/mathlib/blob/lean-3.44.1/src/ring_theory/witt_vector/mul_coeff.lean#L275):
+
+---
+**Multiplication lemma.**
+For every positive $n \in \mathbb{N}$, there is a function $f_n : k^{n} \times k^{n} \to k$ such that for every $x, y \in \mathbb{W}(k)$, the $n$-th coefficient of $x \cdot y$ is given by 
+
+$$ x_{n}y_0^{p^{n}} + y_{n}x_0 ^{p^{n}} + f_n(x_0, \ldots, x_{n-1}, y_0, \ldots, y_{n-1}).  $$
+
+---
```
This might work well as a blockquote (`>` at the start of each line)

hrmacbeth replied:
I prefer it with our ad-hoc horizontal-line formatting, because with the indentation it's not visually clear that the displayed equation is part of the lemma.  OK?

eric-wieser commented on posts/formalizing-isocrystals-in-lean.md:
I think it would be wise to use a permalink here that includes a commit hash, just in case we ever patch old lean branches

eric-wieser commented on posts/formalizing-isocrystals-in-lean.md:
```diff
@@ -0,0 +1,211 @@
+type: text
+---
+
+Last year, there was a [big mathlib refactor](https://leanprover-community.github.io/blog/posts/semilinear-maps/) to replace linear maps throughout the library with  *semilinear maps*,
```
```suggestion
Last year, there was a [big mathlib refactor](link://filename/semilinear-maps.md) to replace linear maps throughout the library with  *semilinear maps*,
```
which ensures that the links don't break if we rename the md (which I assume would give a build error) or move the blog root url.

robertylewis replied:
Local testing implies it needs the full path:
```suggestion
Last year, there was a [big mathlib refactor](link://filename/posts/semilinear-maps.md) to replace linear maps throughout the library with  *semilinear maps*,
```

robertylewis commented on posts/formalizing-isocrystals-in-lean.md:
```diff
@@ -0,0 +1,211 @@
+
+
+So, for the fraction field $K$ of the Witt vectors of an algebraically closed field $k$, what is this multiplicative subgroup
+$ \\{ \varphi(b) / b : b \in K^\* \\}$
```
For whatever reason, this formats correctly with the `\\` and not with the `\`.

robertylewis replied:
@faenuccio when I build the blog locally I see this with the `\\`:
![image](https://user-images.githubusercontent.com/4967469/182179839-e7a1dafa-e7d7-44e3-ad83-8bc73ffacd42.png)

And this with `\`:
![image](https://user-images.githubusercontent.com/4967469/182180076-57e075d7-e7d1-4497-957e-78cc12456afe.png)

faenuccio commented:
Not for me...



--
Filippo A. E. Nuccio Mortarino Majno di Capriglio
Maître de Conférences
Univ de Lyon, Université Jean Monnet
Institut Camille Jordan et IUT
23, rue du docteur Paul Michelon
42023 Saint-Étienne - France

Bureau : B 105 B
Tel : +33 (0) 4 77 48 15 69
E-mail: ***@***.***
web: http://perso.univ-st-etienne.fr/nf51454h/

------ Original Message ------
From "Rob Lewis" ***@***.***>
To "leanprover-community/blog" ***@***.***>
Cc "Filippo A. E. Nuccio" ***@***.***>; "Comment" 
***@***.***>
Date 01/08/2022 16:46:29
Subject Re: [leanprover-community/blog] An isocrystal story (PR #47)

>
***@***.*** commented on this pull request.
>
>--------------------------------------------------------------------------------
>In posts/formalizing-isocrystals-in-lean.md 
><https://github.com/leanprover-community/blog/pull/47#discussion_r934615036>:
>
> > +There is a natural notion of equivalence on these objects: $(V_1, f_1)$ and $(V_2, f_2)$ are equivalent, if there exists a (linear) isomorphism $g:V_1\to V_2$, which is equivariant in the sense that $g \circ f_1=f_2 \circ g$.   In the case of one-dimensional objects over $\mathbb{C}$, this agrees well with our geometric intuition for when two orientation-reversing similarities of planes with marked points are "the same."
>+
>+It's easy to cook up a large family of such objects with $V$ taken to be the field $F$ itself, considered as a one-dimensional vector space over itself:  take the $\sigma$-semilinear automorphism to be $x \mapsto a \sigma(x)$ for any $a\ne 0$ in $F$.  It's also fairly easy to see that, up to the sense of equivalence described above, this covers all possible one-dimensional objects. <!-- say any more? -->
>+
>+But this is not the interesting part of the story.  Some choices of $a\in F$ lead to equivalent objects, and the question is, which? For example, in the case of  $\mathbb{C}$, the equivalent transformations $f_1(z)=a_1\overline{z}$, $f_2(z)=a_2\overline{z}$ are precisely those for which $|a_1|=|a_2|$.  That is, for a given $r\in\mathbb{R}^+$, all transformations $f(z)=a\overline{z}$ with $|a|=r$ are equivalent to $f(z)=r\overline{z}$ itself, geometrically corresponding to the same transformation of a "rescaling by $r$" about the marked point of a reflection in a line through that marked point.
>+
>+![Rescale and reflect](/images/isocrystal-geo.png)
>+
>+In general, the characterization of equivalence is the following: two $\sigma$-semilinear transformations $f_1(x)=a_1\sigma(x)$, $f_2(x)=a_2\sigma(x)$ of $F$ are equivalent if there exists a nonzero element $b\in F$ such that $a_1/a_2 = \sigma(b)/b$.  The set of elements of $F$ of the form $\sigma(b)/b$ forms a multiplicative subgroup of $F^*:=F\setminus\\{0\\}$. In the case of conjugation on $\mathbb{C}$ this subgroup is the unit circle.  And classifiying the equivalence classes of one-dimensional objects reduces to finding a family of canonical representatives of the cosets of this subgroup  -- in the case of $\mathbb{C}$, each coset of the unit circle action has a unique representative $r \in\mathbb{R}^+$.
>+
>+
>+# Frobenius-semilinear automorphisms of the fraction field of the Witt vectors
>+
>+
>+So, for the fraction field $K$ of the Witt vectors of an algebraically closed field $k$, what is this multiplicative subgroup
>+$ \\{ \varphi(b) / b : b \in K^\* \\}$
>
>For whatever reason, this formats correctly with the \\ and not with 
>the \.
>
>—
>Reply to this email directly, view it on GitHub 
><https://github.com/leanprover-community/blog/pull/47#discussion_r934615036>, 
>or unsubscribe 
><https://github.com/notifications/unsubscribe-auth/APQQWUEVQY2DWQUYANKFKUTVW7PMLANCNFSM543QSU5Q>.
>You are receiving this because you commented.Message ID: 
***@***.***>
>

robertylewis commented:
I've pushed a commit that I think addresses the remaining comments. @kbuzzard in regard to your suggestions, we've added a footnote quoting you directly :smile: this discussion seemed a little out of place coming from our perspective -- I still haven't fully processed what you mean myself. But it's an interesting discussion nonetheless and we've referenced the comments on this PR.

adamtopaz commented:
Another nice analogy that might be worth mentioning is that both the unit circle in $$\mathbb{C}$$ and the units $$W(k)^\times$$ in $$K^\times$$ can be computed as the set of elements of absolute value $$1$$.

hrmacbeth commented on posts/formalizing-isocrystals-in-lean.md:
```diff
@@ -0,0 +1,218 @@
+Recall that a Witt vector $x$ over $k$ is given by a sequence $(\ldots, x_1, x_0)$ of elements of $k$. (Witt vectors are traditionally written as trailing to the left, to be suggestive of how we write numbers in decimal notation from largest place value to smallest.)
+The $p$-typical Witt vector $(\ldots, 0,1,0)$ is referred to in $\mathbb{W}(k)$ as $p$ (it actually ends up being the sum of $p$ copies of $1=(\ldots, 0,0, 1)$, under the crazy Witt vector addition).  Multiplication (in the crazy Witt vector sense) by $p$ 
+[sends any Witt vector](https://github.com/leanprover-community/mathlib/blob/25706131162465df8e6daecec75505f782fd428c/src/ring_theory/witt_vector/identities.lean#L37)
+to a Witt vector with $0$ in the first position, and more generally multiplication by $p^m$ for any $m\in \mathbb{N}$ sends any Witt vector to a Witt vector with $m$ leading zeros. In fact, every nonzero Witt vector is [of the form](https://github.com/leanprover-community/mathlib/blob/25706131162465df8e6daecec75505f782fd428c/src/ring_theory/witt_vector/discrete_valuation_ring.lean#L117)  $p^m w$  ...
```
@jcommelin Now that we wrote Witt vectors as trailing to the left, should we be changing the positional terminology?  For example, the reference here to 
> a Witt vector with $0$ in the first position

and a few paragraphs later to
> first $n$ coefficients

Is "first" still ok, or should we say "last coefficient", "lowest coefficient", ...?

riccardobrasca commented on posts/formalizing-isocrystals-in-lean.md:
```diff
@@ -0,0 +1,218 @@
+type: text
+---
+
+Last year, there was a [big mathlib refactor](link://filename/posts/semilinear-maps.md) to replace linear maps throughout the library with  *semilinear maps*,
```
In https://github.com/leanprover-community/blog/blob/isocrystals/posts/formalizing-isocrystals-in-lean.md the hyperlink is not showed.

robertylewis replied:
This syntax is specific to Nikola; it works when the blog is built!

riccardobrasca commented on posts/formalizing-isocrystals-in-lean.md:
```diff
@@ -0,0 +1,218 @@
+structure (dependent on $p$) to make this set into a commutative ring.  The canonical example is the $p$-adic integers, which are the Witt vectors
+of $\mathbb{Z}/p\mathbb{Z}$.
+
+A ring $R$ of characteristic $p$ has an endomorphism, constructed by sending each element $x$ to $x ^ p$.  $R$ is called *perfect* if this
```
```suggestion
A ring $R$ of characteristic $p$ has an endomorphism, constructed by sending each element $x$ to $x ^ p$.  We say that $R$ is *perfect* if this
```
This is just to avoid ending a sentence with a symbol and starting the next one with another symbol.

riccardobrasca commented on posts/formalizing-isocrystals-in-lean.md:
```diff
@@ -0,0 +1,218 @@
+want to consider![^1]
+
+The fundamental result here is a classification theorem for
+Frobenius-semilinear automorphisms of finite-dimensional vector spaces over $K$, the field of fractions of $\mathbb{W}(k)$, for $k$ an algebraically closed field (and hence a perfect integral domain).  Such an object (a finite-dimensional 
```
```suggestion
Frobenius-semilinear automorphisms of finite-dimensional vector spaces over $K$, the field of fractions of $\mathbb{W}(k)$, for $k$ an algebraically closed field of characteristic $p$ (and hence a perfect integral domain).  Such an object (a finite-dimensional 
```

riccardobrasca commented on posts/formalizing-isocrystals-in-lean.md:
```diff
@@ -0,0 +1,218 @@
+
+But to conclude the main discussion, our theorem is that every coset of the special subgroup has a representative of the form $p^m$ for some $m\in\mathbb{Z}$.  This integer $m$ is called the *slope* of the 
+[associated isocrystal](https://github.com/leanprover-community/mathlib/blob/25706131162465df8e6daecec75505f782fd428c/src/ring_theory/witt_vector/isocrystal.lean#L150):
+$K$ itself (considered as a one-dimensional vector space over itself), equipped with the Frobenius-semilinear automorphism sending $x\in K$ to $p^m\varphi(x)$.  All one-dimensional isocrystals are equivalent to an isocrystal of this form. 
```
I find this sentence a little hard to follow. Can you be a little more explicit? Saying that you are building a "standard" isocrystal of slope $m$.

robertylewis replied:
@riccardobrasca I've reworded this paragraph:

> But to conclude the main discussion, our theorem is that every coset of the special subgroup has a representative of the form $p^m$ for some $m\in\mathbb{Z}$.
> We call $K$ (considered as a one-dimensional vector space over itself), equipped with the Frobenius-semilinear automorphism sending $x\in K$ to $p^m\varphi(x)$,
> the [standard isocrystal of slope $m$](https://github.com/leanprover-community/mathlib/blob/25706131162465df8e6daecec75505f782fd428c/src/ring_theory/witt_vector/isocrystal.lean#L150).
> All one-dimensional isocrystals are equivalent to one of these standard isocrystals. 

Does this sound better and accurate? cc @hrmacbeth

hrmacbeth replied:
Looks good to me!

riccardobrasca reviewed (approved):
I don't remember: the proof that `W(k)` is a DVR originated in this work? If this was the case it seems something worth to be mentioned.

robertylewis commented:
(The time was also updated in the last commit)
